Introduce Bad Request Handler Rack Middleware

Math3v opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - Read the DEV Contributing Guide and the Code of Conduct
     - Provided tests for your changes
     - Used descriptive commit messages
     - Updated any relevant documentation and added any necessary screenshots
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Hi all, I'm new to this project and I tried to fix https://github.com/thepracticaldev/dev.to/issues/6245.
My PR introduces a generic rack middleware to rescue bad requests before they happen and gracefully return 400 with a description of the exception. I'm definitely open to any improvements and suggestions.

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/issues/6245

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![Shield](https://media.giphy.com/media/l0HlymZ7Jv6JoiYjC/giphy.gif)

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=6248) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=6248) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=6248) it.</sub>

mstruve commented on lib/middleware/handle_bad_request.rb:
```diff
@@ -0,0 +1,13 @@
+
+    def call(env)
+      @app.call(env)
+    rescue ActionController::BadRequest => e
```
I like this approach but I am a little confused as to how it works. [The Honeybadger](https://app.honeybadger.io/fault/66984/2eb393354a8c34bb594b71ba905d2b69) shows us getting an `ArgumentError` from the code rather than an `ActionController::BadRequest`. If we got a BadRequest error Honeybadger [would actually ignore it itself](https://docs.honeybadger.io/lib/ruby/getting-started/ignoring-errors.html#ignore-by-class) so we shouldn't technically even need Middleware like this. 

Were you able to write a failing spec and get an ArgumentError before implementing this solution?

Math3v replied:
@mstruve Hi, thanks for taking a look!

Yes, the test included was written first and it got me all the way to handling this at the middleware level as the issue does not reach the controller at hand.

I'm attaching a screenshot from my local. You can see there that the error is being caught and handled by the middleware. Also, I discovered and fixed one issue as I learned the `body` needs to respond to `.each`, so it needs to be an array. See: http://programmerscando.com/rails-and-body_proxy-error-nomethoderror/

Screenshot: 
<img width="1168" alt="Screenshot 2020-02-23 at 15 05 05" src="https://user-images.githubusercontent.com/3818442/75113465-7777d000-564e-11ea-9575-78c857f6ae99.png">

rhymes commented:
I agree, I don't think you're handling the correct error @Math3v :)

I was able to reduce it to:

```ruby
> ActiveRecord::Base.connection.quote_string("\0string")
ArgumentError: string contains null byte
from /Users/rhymes/.rvm/gems/ruby-2.6.5/gems/activerecord-5.2.4.1/lib/active_record/connection_adapters/postgresql/quoting.rb:21:in `escape_string'
```

See also:

```ruby
[28] pry(main)> Article.find_by(path: "/username/\0slug")
  Article Load (1.6ms)  SELECT  "articles".* FROM "articles" WHERE "articles"."path" = $1 LIMIT $2  [["path", "/username/\u0000slug"], ["LIMIT", 1]]
ArgumentError: string contains null byte
from /Users/rhymes/.rvm/gems/ruby-2.6.5/gems/activerecord-5.2.4.1/lib/active_record/connection_adapters/postgresql_adapter.rb:622:in `exec_prepared'
```

I'm not sure we can use a catch all middleware to rescue ArgumentError in this case because it might end up hiding future issues, that we'd like to actually find out about. I'm a bit unsure about how to reproduced it in a controller test though, because if I put the null byte in a requests spec I end up triggering BAD URI, as the URI is parsed by rack test:

```ruby
  1) StoriesShow handles invalid slug characters
     Failure/Error: get "/devteam/call-for-contributions-move-all-\0%delayedjobs-to-sidekiq-246k"

     URI::InvalidURIError:
       bad URI(is not URI?): "http://www.example.com:80/devteam/call-for-contributions-move-all-\u0000%delayedjobs-to-sidekiq-246k"
     # /Users/rhymes/.rvm/gems/ruby-2.6.5/gems/rack-test-1.1.0/lib/rack/test.rb:214:in `parse_uri'
     # /Users/rhymes/.rvm/gems/ruby-2.6.5/gems/rack-test-1.1.0/lib/rack/test.rb:117:in `request'
     # ./spec/requests/stories_show_spec.rb:125:in `block (2 levels) in <top (required)>'
```

Maybe you can just use `rescue_from ArgumentError`, like we do here:

https://github.com/thepracticaldev/dev.to/blob/98a2aaf0ce6a1e1a1ac6529f417ff6827f83602d/app/controllers/application_controller.rb#L8

Maybe you can write a test that mocks a method called in that part of `StoriesController`, raises `ArgumentError` and verify it's caught by `rescue_from`

Let me know if that makes sense!

Math3v commented:
@rhymes I addressed your comment and I agree that handling it locally keeps the visibility to other errors. However, now I cannot reproduce the issue locally as I'm still getting 500. I'm not sure what was the real input as it got through the query parsing all the way to `find_by`.

mstruve commented on app/controllers/stories_controller.rb:
```diff
@@ -14,6 +14,8 @@ class StoriesController < ApplicationController
   before_action :authenticate_user!, except: %i[index search show]
   before_action :set_cache_control_headers, only: %i[index search show]
 
+  rescue_from ArgumentError, with: :bad_request
```
Love that we are able to keep this contained to this single controller like this! I'm betting we find more uses for this statement in a couple of other controllers as well 😉
